File: learning_ros2/ros1_to_ros2_migration/src/ros1_ws/ros1_app_py/scripts/number_counter.py
#!/usr/bin/env python3

import rospy
from std_msgs.msg import Int64
from my_robot_msgs.srv import ResetCounter

counter = 0


def callback_number(msg):
    global counter
    counter += msg.data
    rospy.loginfo("Counter: " + str(counter))


def callback_reset_counter(req):
    if req.reset_value >= 0:
        global counter
        counter = req.reset_value
        rospy.loginfo("Counter: " + str(counter))
        return True
    return False


if __name__ == '__main__':
    rospy.init_node('number_counter')

    sub = rospy.Subscriber("number", Int64, callback_number)

    reset_service = rospy.Service(
        "reset_counter", ResetCounter, callback_reset_counter)

    rospy.loginfo("Number counter has been started.")

    rospy.spin()

File: learning_ros2/ros1_to_ros2_migration/src/ros1_ws/ros1_app_cpp/src/number_counter.cpp
#include <ros/ros.h>
#include <std_msgs/Int64.h>
#include <my_robot_msgs/ResetCounter.h>

int counter = 0;

void callback_number(const std_msgs::Int64 &msg)
{
    counter += msg.data;
    ROS_INFO("Counter: %d", counter);
}

bool callback_reset_counter(my_robot_msgs::ResetCounter::Request &req,
                            my_robot_msgs::ResetCounter::Response &res)
{
    if (req.reset_value >= 0)
    {
        counter = req.reset_value;
        res.success = true;
        ROS_INFO("Counter: %d", counter);
    }
    else
    {
        res.success = false;
    }

    return true;
}

int main(int argc, char **argv)
{
    ros::init(argc, argv, "number_counter");
    ros::NodeHandle nh;

    ros::Subscriber sub = nh.subscribe("number", 1000, callback_number);

    ros::ServiceServer reset_service = nh.advertiseService("reset_counter",
                                                           callback_reset_counter);

    ROS_INFO("Number counter has been started.");

    ros::spin();
    return 0;
}
